agent/agent.go

```diff
@@ -379,6 +335,93 @@ func (a *Agent) loop(ctx context.Context) error {
 	}
 }
 
+// ProcessNextResponse contains information for what should be done after processing next task
+type ProcessNextResponse struct {
+	ShouldExit       bool
+	ShouldContinue   bool
+	ShouldResetTimer bool
+	needPostGroup    bool
+	tc               *taskContext
+}
+
+// ProcessNextTask takes in the NextTaskResponse and processes it.
+func (a *Agent) ProcessNextTask(nt *apimodels.NextTaskResponse, ctx context.Context, tc *taskContext, needPostGroup bool) (ProcessNextResponse, error) {
+	if nt.ShouldExit {
+		grip.Notice("Next task response indicates agent should exit.")
+		return ProcessNextResponse{ShouldExit: true}, nil
+	}
+	// if the host's current task group is finished we teardown
+	if nt.ShouldTeardownGroup {
+		a.runPostGroupCommands(ctx, tc)
+		return ProcessNextResponse{
+			ShouldContinue:   true,
+			ShouldResetTimer: true,
+			// Running the post group commands implies exiting the group, so
+			// destroy prior task information.
+			tc:            &taskContext{},
+			needPostGroup: false,
+		}, nil
+	}
+	if nt.TaskId != "" {
+		if nt.TaskSecret == "" {
+			grip.Critical(message.Fields{
+				"message": "task response missing secret",
+				"task":    tc.task.ID,
+			})
+			return ProcessNextResponse{
+				ShouldContinue:   true,
+				ShouldResetTimer: true,
+			}, nil
+		}
+		prevLogger := tc.logger
+		tc = a.prepareNextTask(ctx, nt, tc)
+		if prevLogger != nil {
+			grip.Error(errors.Wrap(prevLogger.Close(), "closing the previous logger producer"))
+		}
+
+		grip.Error(message.WrapError(a.fetchProjectConfig(ctx, tc), message.Fields{
+			"message": "error fetching project config; will attempt at a later point",
+			"task":    tc.task.ID,
+		}))
+
+		a.jasper.Clear(ctx)
+		tc.jasper = a.jasper
+		shouldExit, err := a.runTask(ctx, tc)
+		if err != nil {
+			grip.Critical(message.WrapError(err, message.Fields{
+				"message": "error running task",
+				"task":    tc.task.ID,
+			}))
+			return ProcessNextResponse{
+				ShouldContinue:   true,
+				ShouldResetTimer: true,
+				tc:               tc,
+			}, nil
+		}
+		if shouldExit {
+			return ProcessNextResponse{
+				ShouldExit: true,
+				tc:         tc,
+			}, nil
+		}
+		return ProcessNextResponse{
+			ShouldContinue:   true,
+			ShouldResetTimer: true,
+			tc:               tc,
+			needPostGroup:    true,
+		}, nil
+	} else if needPostGroup {
+		a.runPostGroupCommands(ctx, tc)
+		return ProcessNextResponse{
+			needPostGroup: false,
+			// Running the post group commands implies exiting the group, so
+			// destroy prior task information.
+			tc: &taskContext{},
+		}, nil
+	}
+	return ProcessNextResponse{}, nil
```

Comment:
It's generally preferrable to have your common path (i.e. having a task to run) to be the last return in the function rather than an if case in an `if else if`. Can we restructure the if-elseif and final return so that it returns early if there's no new task to run? Something like this:
```go

if nt.TaskId == "" && needPostGroup {
    // No task but need to run post group, so run it, return, and sleep.
}
if nt.TaskId == "" {
    // No task to run and no post group to run, just return and sleep.
    return processNextResponse{
        noNewTaskToRun: true,
    }
}

// Task ID was returned, so we should actulaly run it.
a.prepareNextTask(...)
a.runTask(...)

return processNextTaskResponse{
    tc: tc,
    // Other fields...
}
```